src/main/java/net/neoforged/neoforge/registries/datamaps/DataMapType.java

```diff
@@ -0,0 +1,169 @@
+ *
+ * <p>
+ * Data maps have the ability of handling conflicts between datapacks that attach an object to the same registry object.
+ * Using {@link #merger() mergers}, {@linkplain DataMapValueMerger#listMerger() collection-based} data maps can, as such, merge values provided by multiple packs
+ * in the same collection. <br>
+ * The {@link DataMapValueMerger#defaultMerger() default merge} will however have the overriding behaviour of "last come wins", similar to recipes.
+ *
+ * <p>
+ * Both datapack registries and normal, built-in registries support data maps.
+ *
+ * @param registryKey   the ID of the registry this data map is for
+ * @param id            the ID of the data map
+ * @param codec         the codec used to decode and encode the values to and from JSON
+ * @param networkCodec  an optional codec that is used to sync the data map to clients
+ * @param mandatorySync if {@code true}, this data map must be present on the client
+ * @param remover       a remover used to remove specific values
+ * @param merger        a merger that merges conflicting values
+ * @param <T>           the type of the data map
+ * @param <R>           the registry the data map is for
+ * @param <VR>          the type of the remover
+ */
+public record DataMapType<T, R, VR extends DataMapValueRemover<T, R>>(
+        ResourceKey<Registry<R>> registryKey,
+        ResourceLocation id,
+        Codec<T> codec, @Nullable Codec<T> networkCodec,
+        boolean mandatorySync,
+        Codec<VR> remover,
+        DataMapValueMerger<T, R> merger) {
+
+    public DataMapType {
+        Preconditions.checkArgument(networkCodec != null || !mandatorySync, "Mandatory sync cannot be enabled when the attachment isn't synchronized");
+    }
+
+    /**
+     * {@return a data map type builder}
+     *
+     * @param id       the ID of the data map
+     * @param registry the key of the registry the data map is for
+     * @param codec    the codec used to deserialize the values from JSON
+     * @param <T>      the type of the data map
+     * @param <R>      the registry the data is for
+     */
+    public static <T, R> Builder<T, R, DataMapValueRemover.Default<T, R>> builder(ResourceLocation id, ResourceKey<Registry<R>> registry, Codec<T> codec) {
+        return new Builder<>(registry, id, codec).remover(DataMapValueRemover.Default.codec());
+    }
+
+    /**
+     * A builder for {@link DataMapType data map types}.
+     *
+     * @param <T>  the type of the data
+     * @param <R>  the registry the data is for
+     * @param <VR> the type of the remover
+     */
+    public static class Builder<T, R, VR extends DataMapValueRemover<T, R>> {
+        private final ResourceKey<Registry<R>> registryKey;
+        private final ResourceLocation id;
+        private final Codec<T> codec;
+
+        private @Nullable Codec<T> networkCodec;
+        private boolean mandatorySync;
+        private Codec<VR> remover;
```

Comment:
Shouldn't this default to the default remover's codec instead of null?

Reply:
No, the motivation against that is that it's not really generically safe. Instead, the builder factory will set the default remover as that's what the default generic of the default builder is.